sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/UnsafeArrayDataBenchmark.scala

```diff
@@ -0,0 +1,256 @@
+    ------------------------------------------------------------------------------------------------
+    Int                                            281 /  296        597.5           1.7       1.0X
+    Double                                         298 /  301        562.3           1.8       0.9X
+    */
+  }
+
+  def writeUnsafeArray(iters: Int): Unit = {
+    val count = 1024 * 1024 * 16
+
+    val intUnsafeRow = new UnsafeRow(1)
+    val intUnsafeArrayWriter = new UnsafeArrayWriter
+    val intBufferHolder = new BufferHolder(intUnsafeRow, 64)
+    intBufferHolder.reset()
+    intUnsafeArrayWriter.initialize(intBufferHolder, count, 4)
+    val intCursor = intBufferHolder.cursor
+    val writeIntArray = { i: Int =>
+      var n = 0
+      while (n < iters) {
+        intBufferHolder.cursor = intCursor
+        val len = count
+        var i = 0
+        while (i < len) {
+          intUnsafeArrayWriter.write(i, 0.toInt)
+          i += 1
+        }
+        n += 1
+      }
+    }
+
+    val doubleUnsafeRow = new UnsafeRow(1)
+    val doubleUnsafeArrayWriter = new UnsafeArrayWriter
+    val doubleBufferHolder = new BufferHolder(doubleUnsafeRow, 64)
+    doubleBufferHolder.reset()
+    doubleUnsafeArrayWriter.initialize(doubleBufferHolder, count, 8)
+    val doubleCursor = doubleBufferHolder.cursor
+    val writeDoubleArray = { i: Int =>
+      var n = 0
+      while (n < iters) {
+        doubleBufferHolder.cursor = doubleCursor
+        val len = count
+        var i = 0
+        while (i < len) {
+          doubleUnsafeArrayWriter.write(i, 0.toDouble)
+          i += 1
+        }
+        n += 1
+      }
+    }
+
+    val benchmark = new Benchmark("Write UnsafeArrayData", count * iters)
+    benchmark.addCase("Int")(writeIntArray)
+    benchmark.addCase("Double")(writeDoubleArray)
+    benchmark.run
+    /*
+    Java HotSpot(TM) 64-Bit Server VM 1.8.0_92-b14 on Mac OS X 10.10.4
+    Intel(R) Core(TM) i5-5257U CPU @ 2.70GHz
+
+    Write UnsafeArrayData:                   Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
+    ------------------------------------------------------------------------------------------------
+    Int                                             82 /   85       2056.9           0.5       1.0X
+    Double                                         139 /  144       1207.1           0.8       0.6X
+    */
+  }
+
+  def getPrimitiveArray(iters: Int): Unit = {
+    val count = 1024 * 1024 * 12
+
+    val intUnsafeArray = new UnsafeArrayData
+    val intSize = calculateHeaderPortionInBytes(count) + 4 * count
+    val intBuffer = new Array[Byte](intSize)
+    Platform.putInt(intBuffer, Platform.BYTE_ARRAY_OFFSET, count)
+    intUnsafeArray.pointTo(intBuffer, Platform.BYTE_ARRAY_OFFSET, intSize)
+    var intPrimitiveArray: Array[Int] = null
+    val readIntArray = { i: Int =>
+      var n = 0
+      while (n < iters) {
+        intPrimitiveArray = intUnsafeArray.toIntArray
```

Comment:
I'm afraid this will be optimized by JIT. How about use a double and `totalLength += intUnsafeArray.toIntArray.length` here?